File: backend/internal/scheme/publicity_test.go
package scheme

import (
	"testing"
	"time"
)

func TestPublicityWindowBoundaries(t *testing.T) {
	start := time.Date(2026, 9, 10, 8, 0, 0, 0, time.UTC)
	end := start.Add(time.Hour)
	w := DefaultWindow()
	if w.PublicityOpen(start) {
		t.Fatal("publicity must default closed")
	}
	w.Publicity = &PublicityWindow{Open: start, Close: end}
	for _, tc := range []struct {
		at   time.Time
		want bool
	}{
		{start.Add(-time.Nanosecond), false}, {start, true}, {end.Add(-time.Nanosecond), true}, {end, false},
	} {
		if got := w.PublicityOpen(tc.at); got != tc.want {
			t.Fatalf("at %s: got %v, want %v", tc.at, got, tc.want)
		}
	}
	// Publication is read-only and may continue after the term is locked.
	w.Open, w.Close, w.Lockdown = start.Add(-3*time.Hour), start.Add(-2*time.Hour), &start
	if err := ValidateTimeline(w, DefaultCapabilities(), DefaultHonorRoll()); err != nil {
		t.Fatal(err)
	}
	for _, invalid := range []PublicityWindow{{Open: start}, {Close: end}, {Open: start, Close: start}, {Open: end, Close: start}} {
		w.Publicity = &invalid
		if err := ValidateTimeline(w, DefaultCapabilities(), DefaultHonorRoll()); err == nil {
			t.Fatal("invalid publicity window accepted")
		}
	}
}

File: frontend/src/lib/publicity.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import { publicityIsOpen } from './publicity.ts'

test('公示默认关闭，开始时开放，结束时关闭，无效时间不能开放原件', () => {
  const window = { open: '2026-09-10T08:00:00Z', close: '2026-09-10T09:00:00Z' }
  const start = Date.parse(window.open), end = Date.parse(window.close)
  assert.equal(publicityIsOpen(undefined, start), false)
  assert.equal(publicityIsOpen(null, start), false)
  assert.equal(publicityIsOpen(window, start - 1), false)
  assert.equal(publicityIsOpen(window, start), true)
  assert.equal(publicityIsOpen(window, end - 1), true)
  assert.equal(publicityIsOpen(window, end), false)
  assert.equal(publicityIsOpen({ open: 'invalid', close: window.close }, start), false)
})
